Flaky TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOnShutdown

```
$ go test -run=TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOnShutdown ./pkg/ruler/...
--- FAIL: TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOnShutdown (6.12s)
    logger.go:41: 2026-01-21T10:36:28.762403384-05:00 level=info msg="cleaning up mapped rules directory" path=/tmp/TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOn178705797/001
    logger.go:41: 2026-01-21T10:36:28.762788359-05:00 level=info user=user-1 msg="updating rule file" file=/tmp/TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOn178705797/001/user-1/test
    logger.go:41: 2026-01-21T10:36:28.762812491-05:00 level=debug msg="creating rule manager for user" user=user-1
    logger.go:41: 2026-01-21T10:36:28.762893293-05:00 user=user-1 probe="internal message: if you see this, probably log initialization has gone wrong"
    logger.go:41: 2026-01-21T10:36:28.763306709-05:00 level=debug user=user-1 caller=manager.go:325 time=2026-01-21T10:36:28.763286842-05:00 msg="Starting provider" provider=static/0 subs=map[config-0:{}]
    logger.go:41: 2026-01-21T10:36:28.763326254-05:00 level=debug msg="updating rules" user=user-1
    logger.go:41: 2026-01-21T10:36:28.763367627-05:00 level=debug user=user-1 caller=manager.go:365 time=2026-01-21T10:36:28.763360216-05:00 msg="Discoverer channel closed" provider=static/0
    logger.go:41: 2026-01-21T10:36:33.764113695-05:00 level=debug user=user-1 caller=sendloop.go:187 time=2026-01-21T10:36:33.764087902-05:00 msg="Starting send loop" alertmanager=http://127.0.0.1:45485/api/v2/alerts
    logger.go:41: 2026-01-21T10:36:33.774927502-05:00 level=info msg="stopping user managers"
    logger.go:41: 2026-01-21T10:36:33.774950424-05:00 level=debug msg="shutting down user manager" user=user-1
    logger.go:41: 2026-01-21T10:36:33.774964958-05:00 level=debug msg="user manager shut down" user=user-1
    logger.go:41: 2026-01-21T10:36:33.774979261-05:00 level=info msg="all user managers stopped"
    logger.go:41: 2026-01-21T10:36:33.774983814-05:00 level=info msg="stopping user notifiers"
    logger.go:41: 2026-01-21T10:36:33.775006264-05:00 level=info user=user-1 caller=manager.go:321 time=2026-01-21T10:36:33.774997691-05:00 msg="Stopping notification manager..."
    logger.go:41: 2026-01-21T10:36:33.775047696-05:00 level=info user=user-1 msg="notifier discovery manager stopped"
    logger.go:41: 2026-01-21T10:36:33.775061128-05:00 level=debug user=user-1 caller=sendloop.go:126 time=2026-01-21T10:36:33.775029622-05:00 msg="Stopping send loop" alertmanager=http://127.0.0.1:45485/api/v2/alerts
    logger.go:41: 2026-01-21T10:36:33.775076693-05:00 level=info msg="all user notifiers stopped"
    logger.go:41: 2026-01-21T10:36:33.775084206-05:00 level=info msg="cleaning up mapped rules directory" path=/tmp/TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOn178705797/001
    manager_test.go:472: 
                Error Trace:    /home/pillin/programming/git_repos/mimir/pkg/ruler/manager_test.go:472
                Error:          Condition never satisfied
                Test:           TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOnShutdown
                Messages:       gave up waiting for second notification to be sent
FAIL
FAIL    github.com/grafana/mimir/pkg/ruler      6.145s
ok      github.com/grafana/mimir/pkg/ruler/notifier     0.009s [no tests to run]
ok      github.com/grafana/mimir/pkg/ruler/rulespb      0.005s [no tests to run]
ok      github.com/grafana/mimir/pkg/ruler/rulestore    0.015s [no tests to run]
ok      github.com/grafana/mimir/pkg/ruler/rulestore/bucketclient       0.015s [no tests to run]
ok      github.com/grafana/mimir/pkg/ruler/rulestore/local      0.016s [no tests to run]
FAIL
```

--- a/pkg/ruler/manager_test.go
+++ b/pkg/ruler/manager_test.go
@@ -396,6 +396,7 @@ func TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOnShutdown(t 
 
 		b, err := io.ReadAll(r.Body)
 		require.NoError(t, err)
+		defer func() { _ = r.Body.Close() }()
 
 		err = json.Unmarshal(b, &alerts)
 		require.NoError(t, err)
@@ -447,7 +448,7 @@ func TestDefaultMultiTenantManager_WaitsToDrainPendingNotificationsOnShutdown(t 
 		require.FailNow(t, "gave up waiting for first notification request to be sent")
 	}
 
-	// Stop the manager, and queue a second notification once the manager is stopped.
+	// Queue a second notification immediately before the manager is stopped.
 	// This second notification will remain in the queue until we release the first notification's request by closing releaseReceiver below.
 	userManager.onStop = func() {
 		userManager.notifier.Send(&notifier.Alert{Labels: labels.FromStrings(labels.AlertName, "alert-2")})
@@ -595,13 +596,13 @@ func (m *managerMock) Run() {
 	defer m.running.Store(false)
 	m.running.Store(true)
 	<-m.done
-
-	if m.onStop != nil {
-		m.onStop()
-	}
 }
 
 func (m *managerMock) Stop() {
+	if m.onStop != nil {
+		m.onStop()
+	}
+
 	close(m.done)
 }
 

